src/main/java/bio/terra/service/dataset/DatasetService.java

```diff
@@ -439,6 +459,86 @@ public List<BulkLoadHistoryModel> getLoadHistory(
     }
   }
 
+  @FunctionalInterface
+  public interface IamAuthorizedCall {
+    void get() throws IamForbiddenException;
+  }
+
+  /**
+   * Throw if the user cannot access the dataset via SAM permissions.
+   *
+   * @param iamAuthorizedCall throws if dataset inaccessible via SAM permissions
+   */
+  void verifyDatasetAccessible(IamAuthorizedCall iamAuthorizedCall) {
+    try {
+      iamAuthorizedCall.get();
+    } catch (Exception iamEx) {
+      throw new ForbiddenException(
+          "Error accessing dataset: see errorDetails",
+          Collections.singletonList(iamEx.getMessage()));
+    }
+  }
+
+  /** Throw if the user cannot read the dataset. */
+  public void verifyDatasetReadable(UUID datasetId, AuthenticatedUserRequest userReq) {
+    IamAuthorizedCall canRead =
+        () ->
+            iamService.verifyAuthorization(
+                userReq, IamResourceType.DATASET, datasetId.toString(), IamAction.READ_DATA);
+    verifyDatasetAccessible(canRead);
+  }
+
+  public DatasetDataModel retrieveData(
+      AuthenticatedUserRequest userRequest,
+      UUID datasetId,
+      String tableName,
+      int limit,
+      int offset,
+      String sort,
+      SqlSortDirection direction,
+      String filter) {
+    Dataset dataset = retrieve(datasetId);
+
+    DatasetTable table =
+        dataset
+            .getTableByName(tableName)
+            .orElseThrow(
+                () ->
+                    new DatasetDataException(
+                        "No dataset table exists with the name: " + tableName));
+
+    if (!sort.equalsIgnoreCase(PDAO_ROW_ID_COLUMN)) {
+      table
+          .getColumnByName(sort)
+          .orElseThrow(
+              () ->
+                  new SnapshotPreviewException(
+                      "No dataset table column exists with the name: " + sort));
+    }
+
+    var cloudPlatformWrapper = CloudPlatformWrapper.of(dataset.getCloudPlatform());
+
+    if (cloudPlatformWrapper.isGcp()) {
+      try {
+        List<String> columns = datasetTableDao.retrieveColumnNames(table, true);
+        String bqFormattedTableName = PDAO_PREFIX + dataset.getName() + "." + tableName;
+        List<Map<String, Object>> values =
+            BigQueryPdao.getTable(
+                dataset, bqFormattedTableName, columns, limit, offset, sort, direction, filter);
+
+        return new DatasetDataModel().result(List.copyOf(values));
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new DatasetDataException("Error retrieving data for dataset " + dataset.getName(), e);
+      }
+    } else if (cloudPlatformWrapper.isAzure()) {
+      throw new NotImplementedException(
+          "Azure datasets are not yet supported for the preview endpoint");
```

Comment:
should we not call it the preview endpoint?
```suggestion
          "Azure datasets are not yet supported for the data endpoint");
```